micro-rdk/src/esp32/camera.rs

```diff
@@ -1,117 +1,254 @@
+        let jpeg_quality = cfg.get_attribute::<i32>("jpeg_quality").unwrap_or(32);
+        //  If pin_sccb_sda is -1, use the already configured I2C bus by number
+        let sccb_i2c_port = cfg.get_attribute::<i32>("sccb_i2c_port").unwrap_or(0);
+
+        let cam = Self {
             config: camera_config_t {
-                pin_pwdn: -1,
-                pin_reset: -1,
-                pin_xclk: 21,
-                __bindgen_anon_1: camera_config_t__bindgen_ty_1 { pin_sccb_sda: 26 },
-                __bindgen_anon_2: camera_config_t__bindgen_ty_2 { pin_sccb_scl: 27 },
-                pin_d7: 35,
-                pin_d6: 34,
-                pin_d5: 39,
-                pin_d4: 36,
-                pin_d3: 19,
-                pin_d2: 18,
-                pin_d1: 5,
-                pin_d0: 4,
-                pin_vsync: 25,
-                pin_href: 23,
-                pin_pclk: 22,
-                xclk_freq_hz: 20000000,
-                ledc_timer: 1,
-                ledc_channel: 1,
-                pixel_format: 4,
-                frame_size: 4,
-                jpeg_quality: 32,
+                pin_pwdn,
+                pin_reset,
+                pin_xclk,
+                __bindgen_anon_1: camera_config_t__bindgen_ty_1 { pin_sccb_sda },
+                __bindgen_anon_2: camera_config_t__bindgen_ty_2 { pin_sccb_scl },
+                pin_d7,
+                pin_d6,
+                pin_d5,
+                pin_d4,
+                pin_d3,
+                pin_d2,
+                pin_d1,
+                pin_d0,
+                pin_vsync,
+                pin_href,
+                pin_pclk,
+                xclk_freq_hz,
+                ledc_channel,
+                ledc_timer,
+                pixel_format: PixelFormat::JPEG as u32,
+                frame_size,
+                jpeg_quality,
+                // Number of frame buffers to be allocated.
+                // If more than one, then each frame will be acquired (double speed)
+                // when pixel_format == jpeg, fb_count > 1 goes to continuous mode, may need to adjust
+                // xclk_freq_hz down to 10_000_000.
                 fb_count: 1,
                 grab_mode: 0,
                 fb_location: 0,
-                sccb_i2c_port: 0,
+                sccb_i2c_port,
             },
-            last_grab: t.now(),
-        }
+        };
+
+        esp!(unsafe { esp_camera_init(&cam.config) })
+            .map_err(|e| CameraError::InitError(Box::new(e)))?;
+        Ok(Arc::new(Mutex::new(cam)))
     }
-    pub fn setup(&self) -> Result<(), CameraError> {
-        let ret = (unsafe { crate::esp32::esp_idf_svc::sys::esp_camera_init(&self.config) })
-            as crate::esp32::esp_idf_svc::sys::esp_err_t;
-        let ret = crate::esp32::esp_idf_svc::sys::EspError::convert(ret);
-        ret.map_err(|e| CameraError::CameraInitError(e.into()))
+
+    fn get_frame(&mut self) -> Result<Esp32CameraFrameBuffer, CameraError> {
```

Comment:
yeah, this doesn't feel like it needs to be its own function

Reply:
I had this to
1. have self so can be used like `cam.get_frame()` instead of `Esp32Cam::get_frame`
2. to just enforce mutability semantics. While there's nothing in the `Esp32Camera` struct that's being mutated, in reality the underlying frame buffer used by the camera is. 

want me to just make it a `self` method instead?

Reply:
@gvaradarajan you want me to add this to the `get_image` and `render_frame` methods instead?

```rs
    impl Camera for Esp32Camera {
        fn get_image(&mut self, mut buffer: BytesMut) {
        let frame =  Esp32CameraFrameBuffer::get().ok_or(CameraError::FailedToGetImage)?;
        ...
    }
}
```

Feels a bit weird if there's nothing to tie this frame buffer to `self` and not use `self` anywhere in the trait function.

Reply:
@acmorrow turns out it needs to be `&mut self` to use `self.get_frame()` properly otherwise I'm getting deref errors.

you prefer the `&mut self` with `self.get_frame()` or just embedding the `Esp32FrameBuffer` creation directly in the `get_image` and `render_frame` calls?

```rs
cannot move out of `*self` which is behind a mutable reference
   --> micro-rdk/src/esp32/camera.rs:181:21
    |
181 |         let frame = self.get_frame()?;
    |                     ^^^^ ----------- `*self` moved due to this method call                                         
    |                     |
    |                     move occurs because `*self` has type `Esp32Camera`, which does not implement the `Copy` trait
```